### PR TITLE
Roman/chat api

### PR DESCRIPTION
**Pull request by jwayne**
Creating a PR so I can compare with main and leave comments .

### CODE REVIEW

File: backend/api/urls.py [hunk 1]
```diff
@@ -1,8 +1,11 @@
...
 urlpatterns = [
-    path("sources", views.sources),
-    path("summary", views.summary),
-    path("source", views.source),
+    path("sources", sources),
+    path("summary", summary),
+    path("source", source),
+    path("chat", chat.chat, name="chat"),
+    path("chat/start", chat.chat_start, name="chat_start"),
+    path("api/chat_history", chat.chat_history, name="chat_history"),
```
Comment (jwayne) [inline backend/api/urls.py:10]: We should remove `api` from the start of this, since all paths in this file are already prefixed with `api/`, by virtue of the root urls file at `backend/server/urls.py`.

Comment (jwayne) [inline backend/api/urls.py:10]: I suggest renaming as follows for a more standard API format. This generally makes it easier for other devs to understand:
* `chat` -> `chat/<chat_id>`
    * This puts the ID in the URL. Django lets you do this as follows: https://stackoverflow.com/a/53983666/1232944
    * Then, change the view function to accept an additional argument `chat_id`
* `chat/start` -> `chat`
* `api/chat_history` -> `chat/<chat_id>`
    * This is the same as #1, but we'll distinguish the two by checking the method
    * Thus you can just remove this line

Standard format:
* Create object: `POST /api/object`
* Get object: `GET /api/object/<id>`
* Modify object: `PUT /api/object/<id>` (in our case I think `POST` is also reasonable, since we're not exactly modifying a chat, but rather adding a message to a chat thread)

File: backend/api/views/chat.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    # create a new cache object if it does not yet exist
+    # now get the summary from the cache
+    chat, _ = Chat.objects.get_or_create(chat_id=chat_id)
+    chat.set_top_sources(top_sources)
+    chat.save()
+    return Response({"chat_id": chat_id})
+
+
+@api_view(["POST"])
+def chat(request):
```
Comment (jwayne) [inline backend/api/views/chat.py:47]: As discussed in my comment on `urls.py`, I'd merge this and `chat_history`, then add an if statement to condition on the method.

```
@api_view(["GET", "POST"])
def chat(request, chat_id):
    if request.method == 'GET':
        return chat_get(request, chat_id)
    else:
        return chat_post(request, chat_Id)

File: backend/api/views/chat.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    )
+
+    # Process the user's message and generate Atla's response
+    chat_history = chat.messages.all()
+    top_sources = chat.get_top_sources()
+    atla_message = process_message(user_message, chat_history, top_sources)
+
+    ChatMessage.objects.create(
+        chat=chat, sender="atla", message=atla_message, timestamp=timezone.now()
+    )
```
Comment (jwayne) [inline backend/api/views/chat.py:71]: I suggest creating both objects at the same time using `ChatMessage.objects.bulk_create`, so this way if `process_message` fails for any reason, we don't end up with a single user message without an Atla response

File: backend/api/views/chat.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    chat, _ = Chat.objects.get_or_create(chat_id=chat_id)
+    chat.set_top_sources(top_sources)
+    chat.save()
+    return Response({"chat_id": chat_id})
+
+
+@api_view(["POST"])
+def chat(request):
+    chat_id = request.data.get("chat_id")
+    user_message = request.data.get("user_message")
```
Comment (jwayne) [inline backend/api/views/chat.py:49]: nit: suggest renaming to `message` for simplicity

File: backend/api/views/chat.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+    if not chat_id:
+        return Response("Missing chat_id", status=status.HTTP_400_BAD_REQUEST)
+
+    chat = get_object_or_404(Chat, chat_id=chat_id)
+    chat_history = list(
+        chat.messages.order_by("timestamp").values("sender", "message", "timestamp")
+    )
+
+    return Response({"chat_history": chat_history})
```
Comment (jwayne) [inline backend/api/views/chat.py:88]: nit: suggest renaming to `messages` for simplicity

File: backend/api/views/utils.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+stop_words_german = stopwords.words("german")
+deduplicator = Deduplicator(mode="jaccard", threshold=0.1, stop_words=stop_words_german)
+cohere_filter = CohereFilter()
+reranker = Reranker(filter_=cohere_filter)
+search = SearchPipeline(semantic_search, reranker, deduplicator)
+
+formats = ["%d.%m.%Y", "%d/%m/%Y", "%Y/%m/%d"]
+
+
+def get_or_compute_search_docs(query: str) -> dict:
```
Comment (jwayne) [inline backend/api/views/utils.py:40]: I found this name confusing, suggested rename: `get_relevant_docs`

File: backend/api/views/summary.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+        docs_context_article_keys, summary = reasoner.get_answer(
+            docs=docs_answer,
+            question=query,
+        )
+
+        # cache the summary
+        cache.set(cache_key_summary, {"summary": summary})
+
+        # cache the highlights for each article -> cache key = (query, document key)
```
Comment (jwayne) [inline backend/api/views/summary.py:56]: nit: `cache the highlighted articles for each doc`?

File: backend/core/utils/cache.py [hunk 8]
```diff
@@ -22,3 +22,6 @@ def get_search_docs_key(self, query: str) -> str:
 
     def get_sources_key(self, query: str) -> str:
         return "sources_%s" % re.sub(" +", "_", query.strip().lower())
+
+    def get_chat_key(self, chat_id: str) -> str:
+        return "chat_%s" % chat_id
```
Comment (jwayne) [inline backend/core/utils/cache.py:27]: Do we need this? Seems odd to cache chats, since we store them in the DB anyway

File: backend/api/views/sources.py [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+def parse_date(date_string, formats, default_date="01.01.1900"):
+    for fmt in formats:
+        try:
+            return datetime.datetime.strptime(date_string, fmt)
+        except ValueError:
+            continue
+    return datetime.datetime.strptime(default_date, "%d.%m.%Y")
+
+
+formats = ["%d.%m.%Y", "%d/%m/%Y", "%Y/%m/%d"]
```
Comment (jwayne) [inline backend/api/views/sources.py:24]: nit: this also appears in `views/utils.py`

File: backend/api/views/sources.py [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+
+
+def split_prmary_secondary(docs: dict, n_primary: int, cutoff_score: float) -> list:
```
Comment (jwayne) [inline backend/api/views/sources.py:27]: typo: `primary`

File: backend/api/views/source.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+    formats,
+)
+from django.http import Http404
+from django.conf import settings
+
+# Cache key manager
+cache_key_manager = CacheKeyManager()
+
+
+def get_top_article_keys(query: str, source_id: str) -> list:
```
Comment (jwayne) [inline backend/api/views/source.py:22]: I suggest we rename to `get_highlights_for_doc`, to consolidate the terminology

**Review by jwayne — COMMENTED** [review]
Nice work, this looks quite reasonable! Left some minor comments.

File: backend/api/views/summary.py (hunk 7 above)
Comment (RomanEngeler1805) [reply backend/api/views/summary.py:52]: that's already how it works. Just below we iterate over the highlights (i.e. article id's) of each document and cache them  separately with `cache_key` the query and the key of the document it belongs to.

Or did I misunderstand you here?

**Comment by RomanEngeler1805** [comment]
Implemented backend changes to support `/chat` API:
- separated the api's out into individual files `chat`, `source`, `sources`, `summary`
- `chat` has a `GET` method for the chat history and a POST method to add a new message to the chat
- introduced a `BaseReasoner` class in the` /reasoner` directory and inherited `ChatReasoner` from it. `ChatReasoner` supports streaming via `Generator` object
- introduced `OpenAIToolClient` in `openai_client` to use OpenAI Functions and to stream the answer
- setup `Chat` and `ChatMessage` models to store a chat conversation (Firestore is left for later)

We use the OpenAI Functions to give the model the possibility to use tools, in our case a semantic search call. This removes the need for an extra classifier. However, it introduced some additional complexity when streaming the OpenAI response, as it first needs to be recognised if it directly gives an answer or if it wants to use the tool.
- if it gives the answer, we can just forward the stream
- if it uses the tool, we first have to "assemble" the tool, execute the search, pass the search results to OpenAI and then stream the answer back

File: .dockerignore [hunk 12]
```diff
@@ -17,11 +17,12 @@ db.sqlite3
 **/static
 
 # Project files to ignore
+/backend/data
 /backend/data/raw
 /backend/data/processed_old
 /backend/data/embeddings
-/backend/crawler
-/backend/preprocessing
+/backend/data/gcs_cache
```
Comment (jwayne) [inline .dockerignore]: If we add `/backend/data`, I think we can skip including all subdirs

File: .dockerignore [hunk 13]
```diff
@@ -17,11 +17,12 @@ db.sqlite3
...
 /backend/experimental
 /backend/tests
-/backend/evaluation
\ No newline at end of file
+/backend/evaluation
+/backend/api/migrations
```
Comment (jwayne) [inline .dockerignore:28]: Why are you excluding the migrations folder?

File: .gitignore [hunk 14]
```diff
@@ -1,6 +1,7 @@
 data
 static
 staticfiles
+backend/api/migrations
```
Comment (jwayne) [inline .gitignore:4]: The migrations folder should be included in both Git and Docker, since it contains code necessary for running migrations.

File: backend/api/models.py [hunk 15]
```diff
@@ -3,17 +3,19 @@
 
 
 class Chat(models.Model):
+    id = models.BigAutoField(primary_key=True)  # new line
     chat_id = models.CharField(max_length=36, unique=True)  # chat_id is unique
     top_sources = models.TextField()  # Store top_sources as a comma-separated string
```
Comment (jwayne) [inline backend/api/models.py:8]: What happens if we add more data to our corpus, and source IDs change? It seems that the DB will contain stale IDs that may refer to the wrong docs.

Do we need to implement a robust solution to setting static IDs to prevent this?

File: backend/api/tests/test_chat_api.py [hunk 16]
```diff
@@ -0,0 +1,129 @@
+from django.test import TestCase, Client
+from django.urls import reverse
+from django.utils import timezone
+from ..models import Chat, ChatMessage
+import json
+from rest_framework import status
+import requests
+
+
+class ChatIntegrationTest(TestCase):
```
Comment (jwayne) [inline backend/api/tests/test_chat_api.py:10]: Personal note: I find it counterproductive to write tests at this early stage. For example, if we change from SQL -> NoSQL (firestore), we'll have to completely rewrite the tests, adding unnecessary dev time

As a comparison point, Opal still doesn't have integration tests for any internal API endpoints. (Rather, we started by adding unit tests for particularly critical / complicated parts of the code). Now, that isn't the best pattern to follow, but just giving an example.

File: backend/api/views/chat.py [hunk 17]
```diff
@@ -1,88 +1,317 @@
...
+    if missing_fields:
+        return Response(
+            f"Missing fields: {', '.join(missing_fields)}",
+            status=status.HTTP_400_BAD_REQUEST,
+        )
+
+    # Generate a unique chat_id and save it along with top_sources to the database
+    # Here we assume that you have a way to generate a unique chat_id
+    chat_id = generate_unique_chat_id()
+    chat = create_chat_and_messages(chat_id, field_values)
```
Comment (jwayne) [inline backend/api/views/chat.py:307]: This function seems to assume that `field_values` contains several keys (`atla_message`, `user_message`) that `validate_required_fields` doesn't check for.

Can we modify `validate_required_fields` to check explicitly that every desired key exists?

File: backend/core/reasoner/reasoner.py [hunk 18]
```diff
@@ -1,165 +1,49 @@
...
+from core.reasoner.base import BaseReasoner
 
-KATEGORIE = {
-    "law": "Gesetz",
-    "court": "Gerichtsurteil",
-    "commentaries": "Kommentar",
-}
 
+class Reasoner(BaseReasoner):
+    """The reasoner class."""
```
Comment (jwayne) [inline backend/core/reasoner/reasoner.py:8]: Now that we have `Reasoner` and `ChatReasoner`, can you add a comment to each explaining the difference?

I assume ChatReasoner includes context on past messages, and Reasoner only factors in the query question, but want to make sure

File: backend/core/storage/data_loader.py [hunk 19]
```diff
@@ -0,0 +1,38 @@
+import json
+import logging
+import os
+
+from core.storage.fs import get_fs
+
```
Comment (jwayne) [inline backend/core/storage/data_loader.py]: I wrote this code, but I recommend adding the following lines (in global scope) before pushing:
```
articles = None
docs = None
```

This way, if `load_docs_global` is not run, we don't run into an error that these imports don't exist when calling e.g. `DJANGO_SKIP_DATA_LOAD=True DJANGO_SETTINGS_MODULE=server.settings.development ./manage.py migrate api`

**Review by jwayne — COMMENTED** [review]
Left some comments, my biggest concern is the models being SQL causing deploy issues. Second biggest being that Chat.SourceIds can become stale if we change the data corpus.

Nice idea with using OpenAI functions.

File: .dockerignore (hunk 13 above)
Comment (RomanEngeler1805) [reply .dockerignore:28]: got it, wasn't aware of that

File: backend/api/tests/test_chat_api.py (hunk 16 above)
Comment (RomanEngeler1805) [reply backend/api/tests/test_chat_api.py:10]: generally agree.

the only reason why I have them is because they help me troubleshoot. I let gpt write the tests, so they don't tend to cover edge cases but still help me troubleshoot

feel free to delete them when they don't make sense anymore. personally, when I for example updated the reranker, I just deleted 80% of the tests, as I didn't bother to adapt them.

File: backend/api/views/chat.py (hunk 17 above)
Comment (RomanEngeler1805) [reply backend/api/views/chat.py:307]: not sure if I get your point here.

`validate_required_fields` takes the `request` and `fields` as parameters, the latter containing (`atla_message`, `user_message`)
it then loops through the fields and checks if they are contained in the request. if not, it appends it to `missing_fields`, which is a return value. if `missing_fields` is not empty, we return and don't create a chat.

would you more explicitly check that (`atla_message`, `user_message`) have length > 0?

File: backend/core/storage/data_loader.py (hunk 19 above)
Comment (RomanEngeler1805) [reply backend/core/storage/data_loader.py]: so we would add them in global scope in `load_docs_global` or define them in the `api/apps.py` as global?
